### PR TITLE
Unify completion pos usage, fix presentation compiler crash in interpolation

### PR DESCRIPTION
**Pull request by rochala**
Fixes crash for added tests in `CompletionInterpolatorSuite` along with unification and better naming for completion pos, source pos, and span usage.

### CODE REVIEW

File: presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala [hunk 1]
```diff
@@ -6,52 +6,54 @@ import java.net.URI
...
+  sourceUri: URI
 ):
-
-  def sourcePos: SourcePosition = cursorPos.withSpan(Spans.Span(start, end))
-  def stripSuffixEditRange: l.Range = new l.Range(cursorPos.offsetToPos(start), cursorPos.offsetToPos(end))
-  def toEditRange: l.Range = cursorPos.withStart(start).withEnd(cursorPos.point).toLsp
-
-end CompletionPos
+  def queryEnd: Int = originalCursorPosition.point
+  def point: Int = originalCursorPosition.point
```
Comment (jkciesluk) [inline presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala:28]: Looks like its the same thing?

File: presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala [hunk 2]
```diff
@@ -6,52 +6,54 @@ import java.net.URI
...
 object CompletionPos:
 
   def infer(
-      cursorPos: SourcePosition,
+      sourcePosition: SourcePosition,
       offsetParams: OffsetParams,
       adjustedPath: List[Tree]
   )(using Context): CompletionPos =
-    infer(cursorPos, offsetParams.uri().nn, offsetParams.text().nn, adjustedPath)
+    infer(sourcePosition, offsetParams.uri().nn, String(sourcePosition.source.content()), adjustedPath)
```
Comment (jkciesluk) [inline presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala:40]: Why this change?

File: presentation-compiler/src/main/dotty/tools/pc/completions/ScalaCliCompletions.scala [hunk 3]
```diff
@@ -26,7 +26,7 @@ class ScalaCliCompletions(
 
   def contribute(dependency: String) =
     val completions = coursierComplete.complete(dependency)
-    val (editStart, editEnd) = CoursierComplete.inferEditRange(pos.point, text)
+    val (editStart, editEnd) = CoursierComplete.inferEditRange(pos.end, text)
```
Comment (jkciesluk) [inline presentation-compiler/src/main/dotty/tools/pc/completions/ScalaCliCompletions.scala:29]: Why this change?

File: presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala (hunk 1 above)
Comment (rochala) [reply presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala:28]: Right, this should also be unified

File: presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala (hunk 2 above)
Comment (rochala) [reply presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala:40]: sourcePosition contains artificial "CURSOR", offsetparams.text() is plain sourcecode

File: presentation-compiler/src/main/dotty/tools/pc/completions/ScalaCliCompletions.scala (hunk 3 above)
Comment (rochala) [reply presentation-compiler/src/main/dotty/tools/pc/completions/ScalaCliCompletions.scala:29]: point in synthetic span (and when you modify it it becomes synthetic), points to the start, not end

File: presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala (hunk 2 above)
Comment (jkciesluk) [reply presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala:40]: Actually looks like `text` is unused and we should remove it

File: presentation-compiler/src/main/dotty/tools/pc/completions/Completions.scala [hunk 4]
```diff
@@ -107,7 +106,7 @@ class Completions(
   end includeSymbol
 
   def completions(): (List[CompletionValue], SymbolSearch.Result) =
-    val (advanced, exclusive) = advancedCompletions(path, pos, completionPos)
+    val (advanced, exclusive) = advancedCompletions(path, completionPos.originalCursorPosition, completionPos)
```
Comment (kasiaMarek) [inline presentation-compiler/src/main/dotty/tools/pc/completions/Completions.scala:109]:
```suggestion
    val (advanced, exclusive) = advancedCompletions(path, completionPos)
```
And to avoid too many changes in `advancedCompletions ` just do:
```
val pos = completionPos.originalCursorPosition
```

File: presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala [hunk 5]
```diff
@@ -6,53 +6,45 @@ import java.net.URI
...
   def infer(
-      cursorPos: SourcePosition,
+      sourcePos: SourcePosition,
       offsetParams: OffsetParams,
       adjustedPath: List[Tree]
   )(using Context): CompletionPos =
-    infer(cursorPos, offsetParams.uri().nn, offsetParams.text().nn, adjustedPath)
+    val identEnd = adjustedPath match
+      case (ident: Ident) :: _ if ident.toString.contains(Cursor.value) =>
+        ident.span.end - Cursor.value.length
```
[the remark this answers is not in the document]
Comment (rochala) [reply presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala]: RefTree.name should handle all the cases

[the remark this answers is not in the document]
Comment (kasiaMarek) [reply presentation-compiler/src/main/dotty/tools/pc/completions/CompletionPos.scala]: including `import`, `export`?

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 6]
```diff
@@ -90,29 +90,20 @@ object Completion:
 
     val completionSymbolKind: Mode =
       path match
-        case untpd.Ident(_) :: untpd.Import(_, _) :: _ => Mode.ImportOrExport
-        case untpd.Ident(_) :: (_: untpd.ImportSelector) :: _ => Mode.ImportOrExport
-        case untpd.Literal(Constants.Constant(_: String)) :: _ => Mode.Term // literal completions
+        case untpd.Ident(_) :: (_: untpd.ImportSelector) :: _ => Mode.ImportOrExport // import scala.@@
```
Comment (kasiaMarek) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala:93]: possibly also should be member?

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 7]
```diff
@@ -211,7 +194,6 @@ object Completion:
       case tpd.Select(qual, _) :: _               if qual.typeOpt.hasSimpleKind => completer.selectionCompletions(qual)
       case tpd.Select(qual, _) :: _                                             => Map.empty
       case (tree: tpd.ImportOrExport) :: _                                      => completer.directMemberCompletions(tree.expr)
```
Comment (kasiaMarek) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala:196]: Is this case needed?

File: presentation-compiler/test/dotty/tools/pc/tests/completion/CompletionKeywordSuite.scala [hunk 8]
```diff
@@ -467,26 +464,16 @@ class CompletionKeywordSuite extends BaseCompletionSuite:
...
       """|object A{
          |  def hello(a: String, u@@)
          |}""".stripMargin,
-      """|unsafe - scala.caps
-         |unsafeNulls - scala.runtime.stdLibPatches.language
-         |unused - scala.annotation
-         |unshared - scala.annotation.internal
-         |unspecialized - scala.annotation""".stripMargin,
-      topLines = Some(5)
+      "",
```
Comment (kasiaMarek) [inline presentation-compiler/test/dotty/tools/pc/tests/completion/CompletionKeywordSuite.scala:476]: why did this change?

Comment (rochala) [reply presentation-compiler/test/dotty/tools/pc/tests/completion/CompletionKeywordSuite.scala:476]: these are illegal completions at this position, it expects the name. On the other hand you can add annotations in this position. I'll add tests that ensure that they are present if we start a query with @

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 6 above)
Comment (rochala) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala:93]: As far as I know member completions are fetched from symbol search. They are also source of extension methods outside the scope. ain this scenario we can only complete with direct members

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 7 above)
Comment (rochala) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala:196]: Yes, those are typed Trees, and this is the head of the path. We need it to find package members

**Comment by rochala** [comment]
I also removed 2 files: Custom completions and its Tests. They were added to the compiler for ammonite, but after all they were also never used. They failed the tests, and the reason behind it that the file hasn't been updated for 3 years, and getting it up to date is too much costs / not worth.

File: presentation-compiler/src/main/dotty/tools/pc/completions/Completions.scala [hunk 9]
```diff
@@ -544,14 +542,7 @@ class Completions(
         else false,
       )
       Some(search.searchMethods(query, buildTargetIdentifier, visitor).nn)
-    else
```
Comment (jkciesluk) [inline presentation-compiler/src/main/dotty/tools/pc/completions/Completions.scala:547]: Why was this removed? If I recon corretly, it was used when we couldn't correctly parse the file, so we've had  `Mode.None`?

Or is removing this a fix, so we don't get completions in comments?

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 10]
```diff
@@ -90,29 +90,20 @@ object Completion:
...
+
+        case untpd.Literal(Constants.Constant(_: String)) :: _ => Mode.Term | Mode.Scope // literal completions
         case (ref: untpd.RefTree) :: _ =>
-          if (ref.name.isTermName) Mode.Term
-          else if (ref.name.isTypeName) Mode.Type
-          else Mode.None
+          val maybeSelectMembers = if ref.isInstanceOf[untpd.Select] then Mode.Member else Mode.Scope
 
-        case (sel: untpd.ImportSelector) :: _ =>
-          if sel.imported.span.contains(pos.span) then Mode.ImportOrExport
```
Comment (jkciesluk) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala:102]: Why remove this check?

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 11]
```diff
@@ -211,7 +208,6 @@ object Completion:
       case tpd.Select(qual, _) :: _               if qual.typeOpt.hasSimpleKind => completer.selectionCompletions(qual)
       case tpd.Select(qual, _) :: _                                             => Map.empty
       case (tree: tpd.ImportOrExport) :: _                                      => completer.directMemberCompletions(tree.expr)
-      case (_: untpd.ImportSelector) :: tpd.Import(expr, _) :: _                => completer.directMemberCompletions(expr)
```
Comment (jkciesluk) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala:214]: This was was unused?

**Review by jkciesluk — COMMENTED** [review]
> They were added to the compiler for ammonite, but after all they were also never used. They failed the tests, and the reason behind it that the file hasn't been updated for 3 years, and getting it up to date is too much costs / not worth.

We also have `AmmoniteIvyCompletions` now, which should do the same thing.

**Comment by rochala** [comment]
> > They were added to the compiler for ammonite, but after all they were also never used. They failed the tests, and the reason behind it that the file hasn't been updated for 3 years, and getting it up to date is too much costs / not worth.
> 
> We also have `AmmoniteIvyCompletions` now, which should do the same thing.

This is actually a bit different. They planned to use this in repl completions.

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 11 above)
Comment (rochala) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala:214]: Those trees are typed, they will be matched by case above.

File: presentation-compiler/src/main/dotty/tools/pc/completions/Completions.scala (hunk 9 above)
Comment (rochala) [reply presentation-compiler/src/main/dotty/tools/pc/completions/Completions.scala:547]: If we can't parse the file there are 2 options:
- parser crashed,
- parser recovered from an error, and we couldn't properly identify path.
The former is bigger problem, and this line wouldn't change a thing (it would still result in empty completions).
When latter happened, it is probably because of completely malformed code, like pasted from other language. I don't think this is an argument that will justify such completions. This line provided a lot of additional results in illegal context such as comments or:

```scala
def hello(a: String, u@@)
```
```scala
Option(1) match {
  case ma@@ =>
}
```

When the `CompletionMode.None` was actually derived correctly.
I don't think that this is the case we should think about, and even if someone pastes the code to modify / adjust it for scala, as soon as they start writing proper code completions pos should be derived correctly.

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 10 above)
Comment (rochala) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala:102]: Actually I have to revert it. I was tricked by our tests, as they don't enter completions in this position for some reason. We should probably also add check to not compute any completions when mode is None...

File: presentation-compiler/src/main/dotty/tools/pc/completions/Completions.scala (hunk 9 above)
Comment (jkciesluk) [reply presentation-compiler/src/main/dotty/tools/pc/completions/Completions.scala:547]: Sounds good.

**Review by kasiaMarek — APPROVED** [review]
lgtm

**Review by jkciesluk — APPROVED** [review]
